# **Recursion**

In some situations it can be beneficial for a function to call itself. This can be used in some sorting and search functions (among many other usecases). Recursion is an important concept in programming languages and can lead to elegant (albeit often inefficient) solutions. Consider the factorial calculation one of the previous sessions:


**Recap - Factorial**

By using only elementary operations and a while loop, we can already express some slightly more advanced mathematical operations like a factorial:

    Factorial:
    5! = 5 * 4 * 3 * 2 * 1
    6! = 6 * 5 * 4 * 3 * 2 * 1
    and so on ...
    Exception: 0! = 1

If we want to express the Factorial of a given number **number**, we can make use of a while loop:

In [49]:
def fac(number):
  if number == 0:           #return 1 if number is 0
    return 1
  else:
    i = number-1            #counter variable that starts at a-1
    while i > 0:            #we count backwards until i equals 1
      number = number * i   #overwrite the initial value with its intermediate value
      i -= 1                #substract 1 from the counter variabl
    return number

fac(5)

120

A more general definition of the factorial could look like this:
    Factorial:
    n! = n * (n-1)!
    0! = 1

    Example:
    4! = 4 * 3 * 2 * 1 => 24
    5! = 5 * 4! => 5 * 24 => 120
  
The definition *n! = n * (n-1)!* is the general case, and the definition of *0! = 1* is the base case.

With both, a recursive function for the calculation of a factorial is possible:

In [51]:
def facRec(number):
  if number == 0:
    return 1
  else:
    return number*facRec(number-1)

facRec(5)

120

Compare the amount of required steps for calculating 5! recursively and iteratively:

Recursion:

    1)
    factorialRecursive(5)
    => 5 * factorialRecursive(4)

    2)
    factorialRecursive(4)
    => 4 * factorialRecursive(3)

    3)
    factorialRecursive(3)
    => 3 * factorialRecursive(2)

    4)
    factorialRecursive(2)
    => 2 * factorialRecursive(1)

    5)
    factorialRecursive(1)
    => 1 * factorialRecursive(0)

    6)
    factorialRecursive(0) = 1

At step 6), the recursion has reached an actual number that can be used to resolve all other expressions:

    7)
    factorialRecursive(1) = 1 * 1

    8)
    factorialRecursive(2) = 2 * 1

    9)
    factorialRecursive(3) = 3 * 2

    10)
    factorialRecursive(4) = 4 * 6
    
    11)
    factorialRecursive(5) = 5 * 24


Internally, the Python Interpreter keeps track of the function calls by putting them on a stack, until the base case is reached. The returned value of the base case is then used to resolve the function calls on the stack.

**Step 1:** Stack push fac(5):

Function Call | Value
-|-
fac(5)|5*fac(4)

**Step 2:** Stack push fac(4):

Function Call | Value
-|-
fac(4)|4*fac(3)
fac(5)|5*fac(4)

**Step 3:** Stack push fac(3):

Function Call | Value
-|-
fac(3)|3*fac(2)
fac(4)|4*fac(3)
fac(5)|5*fac(4)

**Step 4:** Stack push fac(2):

Function Call | Value
-|-
fac(2)|2*fac(1)
fac(3)|3*fac(2)
fac(4)|4*fac(3)
fac(5)|5*fac(4)

**Step 5:** Stack push fac(1):

Function Call | Value
-|-
fac(1)|1*fac(0)
fac(2)|2*fac(1)
fac(3)|3*fac(2)
fac(4)|4*fac(3)
fac(5)|5*fac(4)

**Step 6:** Stack push fac(0) => base case reached:
Function Call | Value
-|-
fac(0)|1
fac(1)|1*fac(0)
fac(2)|2*fac(1)
fac(3)|3*fac(2)
fac(4)|4*fac(3)
fac(5)|5*fac(4)

**Step 7:** Stack pop => 1:
Function Call | Value
-|-
fac(1)|1*1
fac(2)|2*fac(1)
fac(3)|3*fac(2)
fac(4)|4*fac(3)
fac(5)|5*fac(4)

**Step 8:** Stack pop => 1:
Function Call | Value
-|-
fac(2)|2*1
fac(3)|3*fac(2)
fac(4)|4*fac(3)
fac(5)|5*fac(4)

**Step 9:** Stack pop => 2:
Function Call | Value
-|-
fac(3)|3*2
fac(4)|4*fac(3)
fac(5)|5*fac(4)

**Step 10:** Stack pop => 6:
Function Call | Value
-|-
fac(4)|4*6
fac(5)|5*fac(4)

**Step 11:** Stack pop => 24:
Function Call | Value
-|-
fac(5)|5*24

**Step 12:** Stack pop => 120:
Function Call | Value
-|-
|

**Stack empty!**

For comparison, here are the individual steps for the iterative factorial:

    fac(5)
    i = 4 #init i as number - 1

    Step 1)
    number = 5 * 4 => 20
    i - 1 => 3

    Step 2)
    number = 20 * 3 => 60
    i - 1 => 2

    Step 3)
    number = 60 * 2 => 120
    i - 1 => 1

    Step 4)
    number = 120 * 1 => 120
    i - 1 => 0

    END CONDITION MET => return 120

# **Listing elements of a nested data structure**
Another example application for recursive algorithms is the traversal of tree-like datastructures. This could be nested lists, directories in an operating system, or html pages.

Consider the following nested list:

In [52]:
nestedList = [5,4,3,[5,1,2,[59,1,25,[5]]],[5,[519,218,[9,5,[515]]]]]

For tasks like determining the number of individual elements in a nested list, or for calculating the total sum of all elements in the list, it is not possible to simply use the sum() or len() functions:

In [53]:
sum(nestedList) #not a supported operation

TypeError: unsupported operand type(s) for +: 'int' and 'list'

In [54]:
len(nestedList) #only lists the elements of the outer list

5

Since the depth of the nesting is not necessarily known, using nested for loops is also not a viable option. Instead, we iterate over the outer list and recursively

In [56]:
def sumElements(someList):
  print (someList)
  nr = 0
  for element in someList:
    if type(element) == type([]): #if the element is another list
      nr += sumElements(element) #recursive call
    else: #otherwise
      nr += element #just adding the element
  return nr

sumElements(nestedList)

[5, 4, 3, [5, 1, 2, [59, 1, 25, [5]]], [5, [519, 218, [9, 5, [515]]]]]
[5, 1, 2, [59, 1, 25, [5]]]
[59, 1, 25, [5]]
[5]
[5, [519, 218, [9, 5, [515]]]]
[519, 218, [9, 5, [515]]]
[9, 5, [515]]
[515]


1381

**Step 1 - Base Level:**

    [5, 4, 3, [5, 1, 2, [59, 1, 25, [5]]], [5, [519, 218, [9, 5, [515]]]]]
    sum = 0

Current | Sum | Action | Value
-|-|-|-
5|0| 0+5|5
4|5| 5+4|9
3|9| 9+3|12
[5, 1, 2, [59, 1, 25, [5]]]|12+|sumElements([5, 1, 2, [59, 1, 25, [5]]])|unknown

**Step 2 - Recursion Level 1:**

    [5, 1, 2, [59, 1, 25, [5]]]
    sum = 12

Current | Sum | Action|Value
-|-|-|-
5|0| 0+5|5
1|5| 5+1|6
2|6| 6+2|8
[59, 1, 25, [5]]|8+|sumElements([59, 1, 25, [5]])|unknown



**Step 3 - Recursion Level 2:**

    [59, 1, 25, [5]]
    sum = 8

Current | Sum | Action| Value
-|-|-|-
59|0| 0+59|59
1|59| 59+1|60
25|60| 60+25|85
[5]|85+|sumElements([5])|unknown


**Step 4 - Recursion Level 3:**

    [5]
    sum = 85

Current Element | Sum | Action|Value
-|-|-|-
5|0| 0+5|5


Base Case Reached!

**Step 5 - Resolution:**

*Call Stack:*

Call|Function Call | Value
-|-|-
a|sumElements([5])|5
b|sumElements([59, 1, 25, a])|90
c|sumElements([5, 1, 2, b)|98
d|sumElements([12, c])|110

**Step 6 - Recursion Level 1:**

    [5, [519, 218, [9, 5, [515]]]]]
    sum = 110
...

**Step X - Resolution:**

=> 1381

# **Testing your code (Part 1)**

To verify the functionality of newly implemented functions methods and classes, is useful to deploy basic testing methods.

Basic tests should verify the program behavior for unexpected/unintended input (negative values, non-fitting data types), for edge cases (like empty lists, maximum/minimum numbers) and in general should verify the correct execution of every control flow condition.

Take the following Linked List as an example:


In [22]:
class ListElement():
  def __init__(self,value):
    self.value = value
    self.next = None

  def __str__(self):
    return str(self.value)

class LinkedList():
  def __init__(self):
    self.first = None

  def __str__(self):
    current = self.first
    strRep = "["
    while current != None:
      strRep += f" {current} "
      current = current.next
    strRep += "]"
    return strRep

  def addElementAtPosition(self,value,position):
    newElement = ListElement(value)
    if self.first == None and position == 0: #if list is empty and position is 0
      self.first = newElement
    elif position == 0: #if list is not empty and position is 0
      newElement.next = self.first
      self.first = newElement
    else: #for all other cases
      currentPos = 0
      current = self.first
      while not current.next == None and currentPos < position-1: #if position > length, element will be added at the end of the list
        current = current.next
        currentPos += 1
      newElement.next = current.next
      current.next = newElement

[ 5  6  7  8 ]
[ 5  6  7  8  TEST ]


To verify the intended behavior of the implementation, the different possible inputs and states should be checked:

- initializing an empty list
- adding an element to an empty list
  - at position 0
  - at position > 0
- adding an element to a non-empty list
  - at position 0
  - at position > 0
  - in-between existing elements (for a list with more than 1 element)
- adding elements of various data types (including None-type)
- testing the string method for empty/non-empty list

In [30]:
#testing for empty list
newList = LinkedList() #init
print (newList)
newList.addElementAtPosition(5,0) #add element at 0

print (newList)

[]
[ 5 ]


In [26]:
#testing for empty list
newList = LinkedList() #init
newList.addElementAtPosition(5,1) #add element at pos>0

print (newList)

AttributeError: 'NoneType' object has no attribute 'next'

Evidently, the test-case above is not covered by the implementation!

The following line checks only for emptyness when the second condition is also met:

`if self.first == None and position == 0: `

This can be easily fixed by ignoring the position argument for an empty list, by changing the line to:

`if self.first == None`

In [27]:
class ListElement():
  def __init__(self,value):
    self.value = value
    self.next = None

  def __str__(self):
    return str(self.value)

class LinkedList():
  def __init__(self):
    self.first = None

  def __str__(self):
    current = self.first
    strRep = "["
    while current != None:
      strRep += f" {current} "
      current = current.next
    strRep += "]"
    return strRep

  def addElementAtPosition(self,value,position):
    newElement = ListElement(value)
    if self.first == None: #if list is empty (ignoring position)
      self.first = newElement
    elif position == 0: #if list is not empty and position is 0
      newElement.next = self.first
      self.first = newElement
    else: #for all other cases
      currentPos = 0
      current = self.first
      while not current.next == None and currentPos < position-1: #if position > length, element will be added at the end of the list
        current = current.next
        currentPos += 1
      newElement.next = current.next
      current.next = newElement

- Empty and Position > 0:

In [28]:
#testing for empty list
newList = LinkedList() #init
newList.addElementAtPosition(5,1) #add element at pos>0

print (newList)

[ 5 ]


- Empty and Position == 0:

In [29]:
#testing for empty list
newList = LinkedList() #init
newList.addElementAtPosition(5,0) #add element at 0

print (newList)

[ 5 ]


- Non-empty and Position == 0:

In [31]:
newList = LinkedList()
newList.addElementAtPosition(5,0)
print (newList)
newList.addElementAtPosition(10,0)
print (newList)

[ 5 ]
[ 10  5 ]


- Non-empty and Position > 0:


In [32]:
newList.addElementAtPosition(15,2)
print (newList)

[ 10  5  15 ]


- In-between elements (for list with at least 2 elements)

In [33]:
newList.addElementAtPosition(20,1)
print (newList)

[ 10  20  5  15 ]


- Non-empty and Position > Number of elements

In [34]:
newList.addElementAtPosition(25,10000)
print (newList)

[ 10  20  5  15  25 ]



- Adding elements of various data types (including None-type)

In [40]:
newList = LinkedList()
newList.addElementAtPosition("Hi",0)
newList.addElementAtPosition([1,2,3,5],0)
newList.addElementAtPosition(True,0)
newList.addElementAtPosition({"a":1000},0)
newList.addElementAtPosition((1,2,3),0)
print (newList)
newList.addElementAtPosition(None,0)
print (newList)

[ (1, 2, 3)  {'a': 1000}  True  [1, 2, 3, 5]  Hi ]
[ None  (1, 2, 3)  {'a': 1000}  True  [1, 2, 3, 5]  Hi ]
[ None  (1, 2, 3)  None  {'a': 1000}  True  [1, 2, 3, 5]  Hi ]
[ None  (1, 2, 3)  None  {'a': 1000}  True  5  [1, 2, 3, 5]  Hi ]


**Testing the factorial functions:**

In [41]:
def facRec(number):
  if number == 0:
    return 1
  else:
    return number*facRec(number-1)

The obvious cases would be:
- number == 0


In [42]:
facRec(0)

1

- number > 0


In [43]:
facRec(1)

1

- number >> 0

In [44]:
facRec(10)

3628800

In [45]:
facRec(100)

93326215443944152681699238856266700490715968264381621468592963895217599993229915608941463976156518286253697920827223758251185210916864000000000000000000000000

In [46]:
facRec(10000)

RecursionError: maximum recursion depth exceeded

- number < 0

In [47]:
facRec(-1)

RecursionError: maximum recursion depth exceeded

- not a number

In [48]:
facRec("Hi")

TypeError: unsupported operand type(s) for -: 'str' and 'int'

To systematically operationalize these tests, there are important concepts which will be the topic later in the semester:

- Unit tests
- Exception handling
- Input verification